### PR TITLE
[SPARK-27748][SS] Kafka consumer/producer password/token redaction.

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

Kafka parameters are logged at several places and the following parameters has to be redacted:
* Delegation token
* `ssl.truststore.password`
* `ssl.keystore.password`
* `ssl.key.password`

This PR contains:
* Spark central redaction framework used to redact passwords (`spark.redaction.regex`)
* Custom redaction added to handle `sasl.jaas.config` (delegation token)
* Redaction code added into consumer/producer code
* Test refactor

## How was this patch tested?

Existing + additional unit tests.

### CODE REVIEW

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala [hunk 1]
```diff
@@ -193,7 +194,7 @@ private[spark] object KafkaTokenUtil extends Logging {
       | debug=${isGlobalKrbDebugEnabled()}
       | useTicketCache=true
       | serviceName="${clusterConf.kerberosServiceName}";
-      """.stripMargin.replace("\n", "")
+      """.stripMargin.replace("\n", "").trim
```
Comment (gaborgsomogyi) [inline external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala]: This is unrelated fix. There were additional spaces at the end of the generated string. This change removes it. Worth to mention the code works both way but thought it's just ugly.

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala [hunk 2]
```diff
@@ -268,8 +270,8 @@ private[spark] object KafkaTokenUtil extends Logging {
       | serviceName="${clusterConf.kerberosServiceName}"
       | username="$username"
       | password="$password";
-      """.stripMargin.replace("\n", "")
-    logDebug(s"Scram JAAS params: ${params.replaceAll("password=\".*\"", "password=\"[hidden]\"")}")
+      """.stripMargin.replace("\n", "").trim
```
Comment (gaborgsomogyi) [inline external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala]: This is unrelated fix. There were additional spaces at the end of the generated string. This change removes it. Worth to mention the code works both way but thought it's just ugly.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.SECRET_REDACTION_PATTERN
+import org.apache.spark.util.Utils.{redact, REDACTION_REPLACEMENT_TEXT}
+
+private[spark] object KafkaRedactionUtil extends Logging {
+  private[spark] def redactParams(params: Seq[(String, Object)]): Seq[(String, Object)] = {
+    val redactionPattern = SparkEnv.get.conf.get(SECRET_REDACTION_PATTERN)
+    params.map { case (key, value) =>
+      if (key.equalsIgnoreCase(SaslConfigs.SASL_JAAS_CONFIG)) {
+        (key, redactJaasParam(value.asInstanceOf[String]))
```
Comment (dongjoon-hyun) [inline external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: Hi, @gaborgsomogyi .
Is this the only reason why we cannot use `Utils.redact` directly?

Comment (gaborgsomogyi) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: @dongjoon-hyun 
Not sure what you mean only reason. The short answer is yes.

A little but more detailed `SaslConfigs.SASL_JAAS_CONFIG` has different format than any other property. A normal property looks like the following:
`Key=ssl.truststore.password, Value=secret`.
`SaslConfigs.SASL_JAAS_CONFIG` however have the following syntax:
`Key=sasl.jaas.config, Value=org.apache.kafka.common.security.scram.ScramLoginModule required tokenauth=true serviceName="kafka" username="admin" password="admin-secret";`
`Utils.redact` makes a malformed and unreadable string out of it.

**Comment by HeartSaVioR** [comment]
I know this patch just apply redaction to all the places which print configuration as of now, but I'm feeling that Kafka module is too verbose on configuration, especially KafkaConfigUpdater. Assuming debug log level is turned on, you would majorly want to track which consumers are being acquired or evicted, or closed. Logging in KafkaConfigUpdater doesn't help either.

Let's revisit the purpose of logging. I guess we are logging to find when consumer/producer is created, and when it's evicted/closed. (either tracking each instance, or counting for each key). Though we use whole configuration as a key of cache, we (as human) cannot check every key-value pairs to find same instance. Some pairs play as a differentiator and we (again, as human) just assume other pairs are not different.

So IMHO we don't even need to print out all configurations. Especially, if we really want to track each instance, explicit ID (as #19096 introduces) would work better and much helpful to debug.

**Comment by gaborgsomogyi** [comment]
> So IMHO we don't even need to print out all configurations.

Without printing out all the configurations I'm wondering how do you think one can find out why a consumer/producer creation fail?

> explicit ID (as #19096 introduces) would work better and much helpful to debug.

I agree, introducing an ID would be enough but only when consumer/producer already created. Such case not necessary to log all configurations. This is the next PR in my queue but didn't want to mix up the 2 changes.

> Kafka module is too verbose on configuration, especially KafkaConfigUpdater.

Yeah, this can be discussed.

**Comment by HeartSaVioR** [comment]
>> So IMHO we don't even need to print out all configurations.

> Without printing out all the configurations I'm wondering how do you think one can find out why a consumer/producer creation fail?

Maybe I was not clear, more clearly I don't think we should print out configurations all the time. Does Kafka module support configuration change in runtime? If not a data source needs to print out these configuration only once, but for every creation/termination of instances.

**Comment by gaborgsomogyi** [comment]
> Maybe I was not clear, more clearly I don't think we should print out configurations all the time.

Same understanding. So this means redaction is needed.

> Does Kafka module support configuration change in runtime?

I don't think so but I'll sit together with Kafka guys, have a deeper understanding and based on that we can go forward. That said my intention is to reduce log size drastically and introducing ID tracking but I suggest to do it in an additional PR to keep the focus.

**Comment by HeartSaVioR** [comment]
I meant Kafka module is `spark-sql-kafka`, as we are concerning about config in spark-sql-kafka. Does we assume the config can be changed after query is being run? I have been simply guessing we don't support it, but the risk of supporting it may be just cache miss, so I'm not sure it is unsupported.

**Comment by gaborgsomogyi** [comment]
From Spark side we're not changing the configuration otherwise caching doesn't make sense.

**Comment by HeartSaVioR** [comment]
As this patch fixes the issue the current code has, IMHO we can go with this patch, and apply the new idea of improvements in another issues/PRs. I can help dealing with applying new idea as well.

**Comment by gaborgsomogyi** [comment]
@dongjoon-hyun do you have any other comments?

**Comment by gaborgsomogyi** [comment]
not much movement, maybe @vanzin ?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala [hunk 4]
```diff
@@ -63,9 +65,12 @@ private[kafka010] object CachedKafkaProducer extends Logging {
       .removalListener(removalListener)
       .build[Seq[(String, Object)], Producer](cacheLoader)
 
-  private def createKafkaProducer(producerConfiguration: ju.Map[String, Object]): Producer = {
-    val kafkaProducer: Producer = new Producer(producerConfiguration)
-    logDebug(s"Created a new instance of KafkaProducer for $producerConfiguration.")
+  private def createKafkaProducer(paramsSeq: Seq[(String, Object)]): Producer = {
+    val kafkaProducer: Producer = new Producer(paramsSeq.map(x => x._1 -> x._2).toMap.asJava)
```
Comment (vanzin) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala]: Why do you need the `.map`? Seems like just `.toMap` would work?

If you really need it, then use `.map { x => ... }`.

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      } else {
+        val (_, newValue) = redact(Some(redactionPattern), Seq((key, value.asInstanceOf[String])))
```
Comment (vanzin) [inline external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: Move `Some(...)` outside the loop.

Also it's a little weird that the seq contains Objects but you're blindly casting them to String. Are you sure that's safe?

**Review by vanzin — COMMENTED** [review]
These logs do seem a bit verbose, but since they're already there...

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala (hunk 4 above)
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala]: Not needed, removed.

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala (hunk 5 above)
Comment (gaborgsomogyi) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: `Some` moved.
The blind cast is a silly problem so covered with test.

File: external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      if (key.equalsIgnoreCase(SaslConfigs.SASL_JAAS_CONFIG)) {
+        (key, redactJaasParam(value.asInstanceOf[String]))
+      } else {
+        value match {
+          case s: String =>
+            val (_, newValue) = redact(redactionPattern, Seq((key, s))).head
+            (key, newValue)
+
+          case _ =>
+            (key, value)
```
Comment (vanzin) [inline external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: Hmm... couldn't the stringified `value` have content you want to redact? Might be better to apply redaction on `value.toString`.

Comment (gaborgsomogyi) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: The type possibilities are the following: `int`, `long`, `short`, `double`, `boolean`, `class`, `list`.
From these only the `list` could be interesting. `list` can contain classes, hostnames or cipher suites. As a result I don't think `toString` has to be redacted but we can punt in to be defensive.

Comment (vanzin) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: Well, redaction is based on the key, and if there's some secret that just happens to have an `int` value, you still don't want that to show up.

Comment (gaborgsomogyi) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: Considering your argument you're right, redacting all other types would be the flawless way.

Redacting other data types, for example `int` I see mainly 2 ways:
* Change the data type to `String` and set the known value `*********(redacted)`. I think it's kind of weird that all of a sudden the parameter type changes. Yeah, in this case would work but still weird.
* Implement the default redacted value for each type. This would increase the complexity and the value is questionable because in case of `int` the user can't really tell 0 is the normal or redacted value. This solution could trick sysadmin/devs by showing invalid values.

Considering the mentioned possibilities + the fact that until now I haven't seen the need for this feature I'm not fully convinced that this has to be added.

Comment (vanzin) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: I really don't understand what you're talking about.

You redact based on the key, or the stringified value. You don't change the type anywhere. This just affects what is printed to the logs.

Comment (gaborgsomogyi) [reply external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaRedactionUtil.scala]: > This just affects what is printed to the logs.

True. I've changed the return type of `redactParams` to make it clear that the value type changes.

**Comment by vanzin** [comment]
merging to master.
